src/persist-client/src/write.rs

```diff
@@ -266,36 +294,72 @@ where
+    ///
+    /// The batch may be empty, which allows for downgrading `upper` to
+    /// communicate progress. It is possible to heartbeat a writer lease by
+    /// calling this with `new_upper` equal to `self.upper()` and an empty
+    /// `updates` (making the call a no-op).
+    ///
+    /// The clunky multi-level Result is to enable more obvious error handling
+    /// in the caller. See <http://sled.rs/errors.html> for details.
+    ///
+    /// SUBTLE! Unlike the other methods on WriteHandle, it is not always safe
+    /// to retry [ExternalError]s in compare_and_append (depends on the usage
+    /// pattern). We should be able to structure timestamp binding, source, and
+    /// sink code so it is always safe to retry [ExternalError]s, but SQL txns
+    /// will have to pass the error back to the user (or risk double committing
+    /// the txn).
+    ///
+    /// TODO: This already retries [mz_persist::location::Determinate] errors,
+    /// so the signature could be changed to only return Indeterminate, but
+    /// leaving it as ExternalError for now to save churn on storage PR rebases.
+    pub async fn compare_and_append_core(
+        &mut self,
+        batch: Batch<K, V, T, D>,
+        expected_upper: Antichain<T>,
+        new_upper: Antichain<T>,
+    ) -> Result<Result<Result<(), Upper<T>>, InvalidUsage<T>>, ExternalError> {
+        trace!(
+            "WriteHandle::compare_and_append_core expected_upper={:?} new_upper={:?}",
+            expected_upper,
+            new_upper
+        );
+
         let lower = expected_upper.clone();
         let upper = new_upper;
         let since = Antichain::from_elem(T::minimum());
         let desc = Description::new(lower, upper, since);
 
-        // TODO: Instead construct a Vec of batches here so it can be bounded
-        // memory usage (if updates is large).
-        let value = match Self::encode_batch(&desc, updates) {
-            Ok(x) => x,
-            Err(err) => return Ok(Err(err)),
-        };
-        let keys = if let Some(value) = value {
-            let key = Uuid::new_v4().to_string();
-            let () = retry_external("compare_and_append::set", || async {
-                // If MultiBlob::set took value as a ref, then we wouldn't have
-                // to clone here.
-                self.blob
-                    .set(
-                        Instant::now() + FOREVER,
-                        &key,
-                        value.clone(),
-                        Atomicity::RequireAtomic,
-                    )
-                    .await
-            })
-            .await;
-            vec![key]
-        } else {
-            vec![]
-        };
+        // TODO(aljoscha): Think hard to make sure these are correct.
+        assert!(PartialOrder::less_equal(batch.desc.lower(), desc.lower()));
+        assert!(PartialOrder::less_equal(desc.upper(), batch.desc.upper()));
+
+        let keys = batch.blob_keys;
 
         loop {
```

Comment:
something I was thinking about after you merged the last PR. I suspect there's some way for us to push some form of this loop into the Machine::compare_and_append call, which I think might make the code a little neater

Reply:
> something I was thinking about after you merged the last PR. I suspect there's some way for us to push some form of this loop into the Machine::compare_and_append call, which I think might make the code a little neater


done!